### ISSUE TITLE
nx affected:e2e is hanging on last app running

### ISSUE THREAD

**Issue by alexw10** [nrwl/nx#8286]
## Current Behavior

Hanging happens it seems to be on the last e2e app that runs
Individually the e2e apps work fine when doing `nx e2e app-name-e2e` for example

## Expected Behavior

Should not hang

Seems so this did not seem to be a problem with Angular 12 and NX 12

## Steps to Reproduce

Yes see below.

I was able to reproduce this, if you pull down master and change in the nx.json change:
```
  "tasksRunnerOptions": {
    "default": {
      "runner": "@nrwl/nx-cloud",
      "options": {
        "accessToken": "YmM1MGMwMTUtNzAyYi00ZjM4LWExNGUtZDM2ZjYyNzNiODAxfHJlYWQ=",
        "cacheableOperations": ["build", "lint", "test"]
      }
    }
  },
```

TO the non NX cloud way of running the tests:
```
  "tasksRunnerOptions": {
    "default": {
      "runner": "@nrwl/workspace/tasks-runners/default",
      "options": {
        "cacheableOperations": ["build", "lint", "test", "e2e"],
      }
    }
  },
```

**It seems to hang... any ideas?**

### Environment

(base) ➜  ~/projects/nx-examples master✘ ✹ npx nx report

>  NX  Report complete - copy this into the issue template

  Node : 16.13.0
  OS   : darwin x64
  yarn : 1.22.17
  
  nx : 13.2.0-rc.4
  @nrwl/angular : 13.2.0-rc.4
  @nrwl/cli : 13.2.0-rc.4
  @nrwl/cypress : 13.2.0-rc.4
  @nrwl/devkit : 13.2.0-rc.4
  @nrwl/eslint-plugin-nx : 13.2.0-rc.4
  @nrwl/express : undefined
  @nrwl/jest : 13.2.0-rc.4
  @nrwl/linter : 13.2.0-rc.4
  @nrwl/nest : undefined
  @nrwl/next : undefined
  @nrwl/node : undefined
  @nrwl/nx-cloud : 12.5.4
  @nrwl/react : 13.2.0-rc.4
  @nrwl/react-native : undefined
  @nrwl/schematics : undefined
  @nrwl/tao : 13.2.0-rc.4
  @nrwl/web : 13.2.0-rc.4
  @nrwl/workspace : 13.2.0-rc.4
  @nrwl/storybook : 13.2.0-rc.4
  @nrwl/gatsby : undefined
  typescript : 4.4.4
  rxjs : 6.5.5
  ---------------------------------------
  Community plugins:
         @angular/animations: 13.0.2
         @angular/common: 13.0.2
         @angular/compiler: 13.0.2
         @angular/core: 13.0.2
         @angular/forms: 13.0.2
         @angular/platform-browser: 13.0.2
         @angular/platform-browser-dynamic: 13.0.2
         @angular/router: 13.0.2
         @ngrx/component-store: 13.0.1
         @ngrx/effects: 13.0.1
         @ngrx/entity: 13.0.1
         @ngrx/router-store: 13.0.1
         @ngrx/store: 13.0.1
         @angular-devkit/build-angular: 13.0.3
         @angular/cli: 13.0.3
         @angular/compiler-cli: 13.0.2
         @angular/language-service: 13.0.2
         @ngrx/store-devtools: 13.0.1

(node:27597) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./" in the "exports" field module resolution of the package at /Users/wintea7/projects/nx-examples/node_modules/tslib/package.json.
Update this package.json to use a subpath pattern like "./*".
(Use `node --trace-deprecation ...` to show where the warning was created)

**Comment by alexw10** [comment]
Is no one else hitting this? :( really causing us not to go to Angular 13/Latest nx

**Comment by Hilligsoe** [comment]
We are running in to what seems to be the same issue with NX hanging after running some of our E2E tests and never finishing after getting `NX SUCCESS` confirmation. 
In our case its (luckily not on all runs) happens after running our shorter/faster component tests. Furthermore, It have been observed that the caching stage for our jest test at times can take up to 8 min. before our pipeline moves on.

I've seen talk about NX potentially losing track of the info it needs to send (This is only theorized), particularly for the jest test, which could be solved by using `--runInBand`, though the author have later removed the issue from stackoverflow.

Not sure if it's exactly the same issue we are facing, but wanted to increase the attention around it.

**Comment by rhmoller** [comment]
We are experiencing something similar where cypress component tests complete and the result is published to nx cloud, but then it hangs until our build pipeline times out.

**Comment by alexw10** [comment]
Yeah I reproduced this in the nx-examples in the description wonder how we can get the attention of this repo where someone might know what to do or have suggestions I am pretty lost on how to fix this.

**Comment by StalkAltan** [comment]
Hi @alexw10, thanks for the reproduction steps. Seems like this issue slipped through the cracks, but we are looking into it now.

**Comment by alexw10** [comment]
@StalkAltan  yay thanks for the update appreciate it!

**Comment by FrozenPandaz** [comment]
Hey all, I can no longer reproduce this in `nx-examples` with `13.4.4`. Could someone please retry this with the latest version?

**Comment by barbados-clemens** [comment]
confirming that I cannot reproduce this on the nx-examples repo either.

**Comment by alexw10** [comment]
@FrozenPandaz @barbados-clemens I pulled down latest master and I am still seeing the issue.
![image](https://user-images.githubusercontent.com/9453636/149584141-0a3c31b0-7a00-4d43-bdfe-a59bc23dc379.png)
Basically it runs the cart e2e and then just doesnt move after that it will just stay like this.
Command I ran is `npm run affected:e2e -- --all`

**Comment by alexw10** [comment]
Oh another note @FrozenPandaz @barbados-clemens  it seems it will work if you run it once and kill it then run it again it will work because it seems it caches it which makes it fully work or something but if you run it with 
`npm run affected:e2e -- --skip-nx-cache --all` it will always freeze up...

**Comment by alexw10** [comment]
So one of our devs seems to have resolved this issue.
He ran both e2e tests separately at same time and noticed port already in use error.
In angular.json he updated the ports to be different for e2e's running and this seems to have resolved the problem...

Not sure if this is an action item for you guys or not but I guess at a minimum we have a work around

**Comment by barbados-clemens** [comment]
Thanks for the extra info. I'll take a deep look at this here soon.

**Comment by dario-digregorio** [comment]
I have this issue with run-many too. I am using `nx run-many --target=build --projects=projectA`. The issue is also happening when using test and e2e as target.

**Comment by barbados-clemens** [comment]
I was finally able to reproduce this by telling nx to run in parallel

```shell
nx run-many --target=e2e --all --skip-nx-cache --parallel --max-parallel 3
```
before all tests looked to be executing one at a time preventing the port collision.

So as mentioned by @alexw10 using a different port for each project should prevent this issue. 

also setting max-parallel to 1 also works with the caveat that only 1 project will run at a time if you do not want to change the ports

```shell
nx run-many --target=e2e --all --skip-nx-cache --max-parallel 1
```
Still looking into why Nx isn't exiting with an error when this port collision happens.

**Comment by barbados-clemens** [comment]
> We have the same issue, it just hangs-on when we using DTE feature.
> 
> [ > NX No new messages received in 60 minutes
> 
> ##[error]Bash exited with code '1'.](url)

have you tried putting each project on a separate port and running the parallelization to 1? `--parallel=1`

**Comment by elvirdolic** [comment]
Here is my command:

`node --max_old_space_size=8192 node_modules/@nrwl/cli/bin/nx.js affected --base=origin/master --head=HEAD --target=e2e --exclude='$(EXCLUDE-E2E)' --parallel=1 --spec=**/integration/**/ci/**/*`

and it still hangs on E2E. I see all tests are run but nothing happens after the last one is run.

The same command is working without DTE feature enabled.

Here is the whole pipeline:

```
    - script: yarn nx-cloud start-ci-run
      displayName: "Start-CI-Run"
    - script: node --max_old_space_size=8192 node_modules/@nrwl/cli/bin/nx.js affected --base=origin/master --head=HEAD --target=build --parallel --max-parallel=4
      displayName: Build
    - script: node --max_old_space_size=8192 node_modules/@nrwl/cli/bin/nx.js affected --base=origin/master --head=HEAD --target=test --parallel --max-parallel=4
      displayName: Test
    - script: node --max_old_space_size=8192 node_modules/@nrwl/cli/bin/nx.js affected --base=origin/master --head=HEAD --target=lint --parallel --max-parallel=4
      displayName: Lint
    - script: node --max_old_space_size=8192 node_modules/@nrwl/cli/bin/nx.js affected --base=origin/master --head=HEAD --target=e2e --exclude='$(EXCLUDE-E2E)' --parallel=1 --spec=**/integration/**/ci/**/*
      displayName: E2E
    - script: yarn nx-cloud stop-all-agents
```

**Comment by barbados-clemens** [comment]
can you try with the --verbose flag as well @elvirdolic?

**Comment by elvirdolic** [comment]
Yes, sure, what information do you would need from the output?

**Comment by barbados-clemens** [comment]
Typically if there was a hidden error it'll show it when using the --verbose flag in my experience. i.e. like a port is already in use or failing to start the dev server for whatever reason. just more of a sanity check to make sure there isn't some error getting missed @elvirdolic

**Comment by fs-innonova** [comment]
I still have the same issue.. I have 3 apps that run on 4200 and I get the same error "address already in use"

**Comment by fkolar** [comment]
I think mine is similar. I just download latest nx, generated plugin setup `create-nx-plugin` tried to run e2e tests and it simply does not finish.


```sh
npx create-nx-plugin my-org --pluginName cli
cd my-org 
nx e2e cli-e2e
```

the e2e tests simply does not finish. it is just hanging.. 

```
 nx --version
14.1.7

node -v
v16.15.0
```

Is there a any recommended node version?

**Comment by barbados-clemens** [comment]
by default, create-nx-plugin is going to be using jest for e2e tests so you're more than likely experiencing something different @fkolar. You can always try using --verbose to see if any extra errors are being thrown that aren't propagating. 
you can also manually run `jest` via the terminal and see if you're getting the same behavior (as the @nrwl/jest plugin _essentially_ just calls jest cli)

**Comment by fkolar** [comment]
@barbados-clemens Thanks for coming back to me. I will try. It seems that there is some problem with project name length. I was just quickly tested a few variations and as soon as folder name is longer, then you can see this problem:

for example this works:

```
npx create-nx-plugin my-org --pluginName cli
npx create-nx-plugin test-wc --pluginName cli
npx create-nx-plugin ttest-wc --pluginName cli
npx create-nx-plugin tttest-wc --pluginName cli
npx create-nx-plugin tttttest-wc --pluginName cli
```

but as soon as you reach certain limits such as :
``
  npx create-nx-plugin t-tttttest-wc --pluginName cli
```
and try to run e2e tests it fails or hangs. but I will try verbose.

**Comment by fkolar** [comment]
So verbose does not print anything

**Comment by mklueh** [comment]
> can you try with the --verbose flag as well @elvirdolic?

facing the same issue with two Nuxt projects, plus that --verbose isn't working https://github.com/nrwl/nx/issues/11091 and I cannot find the cause of my problem for that reason

**Comment by dougludlow** [comment]
I ran into this a couple ways. The first was a test that that had an open handle that I was able to detect thanks to jest: https://jestjs.io/docs/cli#--detectopenhandles. The other was a target that had a `dependsOn` that ran `yarn install` with `nx:run-commands`, but the `cwd` didn't exist so for some reason was just causing `nx affected` to hang without any visible error (even with `--verbose`). The way I tracked that one down was commenting out the target I was running in the `project.json` of each project until I found the offending one that was hanging and then running that individual target with `nx run`. Hope that's helpful for any others who run into this.

**Comment by jfahrenkrug** [comment]
I'm having the same issue when running tests in our NX app on Bitbucket Cloud. It seems to be connected to running tests in parallel. When `max-parallel` is `3`, all the tests run, but the actual command never seems to finish on Bitbucket (works fine when I run it locally). When I set `max-parallel` to `1`, it finishes fine on Bitbucket Cloud. This is the line in question:

```
npx nx affected --target=test --base=origin/master --parallel --max-parallel=3 --ci --code-coverage
```

I've also asked a question on the Bitbucket Forum here:
https://community.atlassian.com/t5/Bitbucket-questions/CI-Pipeline-hangs-when-using-nx-with-max-parallel-higher-than-1/qaq-p/2161467#M86672

**Comment by barbados-clemens** [comment]
anyone have a cloneable repo I can use to look into this? I have yet to be able to reproduce it locally. The original issue is around Cypress so those with a jest issue will need to open an new issue.

**Comment by c01inQvQ** [comment]
I've run into the similar issue reported by @jfahrenkrug.

When running jest tests on CircleCI with

```bash
$ nx affected --target=test --base=$NX_BASE --head=$NX_HEAD
```

Some tests run successfully but the remaining tests hanged without any output logs.
And the 10mins timeout reached with the following error on CI pipeline

```bash
Too long with no output (exceeded 10m0s): context deadline exceeded
```

After delving into the documents/issues, I found the following articles which recommend to use `--runInBand` for Jest and letting Nx handle the degrees of parallelism
- https://nx.dev/packages/jest#performance-in-ci
- https://github.com/nrwl/nx/issues/8964#issuecomment-1043461553

So I tried to modify the command as

```bash
$ nx affected --target=test --base=$NX_BASE --head=$NX_HEAD --runInBand
```

Everything works fine and the performance of the test has indeed improved. Instead of reaching the 10mins timeout, the tests are finished in 1min.

But I still don't know why the difference in performance is so large. Actually, I don't think it's the root cause 😭

**Comment by scott-ut** [comment]
> I've run into the similar issue reported by @jfahrenkrug.
> 
> When running jest tests on CircleCI with
> 
> ```shell
> $ nx affected --target=test --base=$NX_BASE --head=$NX_HEAD
> ```
> 
> Some tests run successfully but the remaining tests hanged without any output logs. And the 10mins timeout reached with the following error on CI pipeline
> 
> ```shell
> Too long with no output (exceeded 10m0s): context deadline exceeded
> ```
> 
> After delving into the documents/issues, I found the following articles which recommend to use `--runInBand` for Jest and letting Nx handle the degrees of parallelism
> 
> * [nx.dev/packages/jest#performance-in-ci](https://nx.dev/packages/jest#performance-in-ci)
> * [Running test in angular is faster for sequential runs than parallel runs #8964 (comment)](https://github.com/nrwl/nx/issues/8964#issuecomment-1043461553)
> 
> So I tried to modify the command as
> 
> ```shell
> $ nx affected --target=test --base=$NX_BASE --head=$NX_HEAD --runInBand
> ```
> 
> Everything works fine and the performance of the test has indeed improved. Instead of reaching the 10mins timeout, the tests are finished in 1min.
> 
> But I still don't know why the difference in performance is so large. Actually, I don't think it's the root cause 😭

I'm seeing the same issue with the build step of my (large, >150 project) monorepo - sometimes after the last project has been built then the command hangs until it is killed by CircleCI.

I've also observed when run locally that it sometimes takes more than 30 seconds to upload to Nx Cloud after a project has been built, and also that I have not been getting cache hits from Nx Cloud when running.

**Comment by barbados-clemens** [comment]
@scott-ut @c01inQvQ this issue is about cypress e2e tests not jest, if there is an issue with jest tests then open a seperate issue. 

I'm going to close this issue as I've not been able to get a reliable reproducible repo and this is initially reported for nx v13 and nx v15 is out. 

If this issue is still relevant open a new issue with a clone-able minimal reproducible so I can take a look. 

Note: getting a port conflict is not an Nx bug since you're telling cypress to spin up a dev server on that same port. We are adding a feature to allow defining ports to be used on the dev server with cypress in this pr: #15486

**Comment by lflo5727** [comment]
@barbados-clemens  I currently don't have a reproducible repo, but I am seeing this with Ember.js tests on NX `15.8.6`.
I verified that tests were not running on the same port. They all finish successfully but hang after the following is printed.
```
> NX Successfully ran target test-ember for 3 projects
```

The tricky part about reproducing it is that it works perfectly fine on my local machine. It seems to hang when being run on our CI (Bamboo)

**Comment by HashemKhalifa** [comment]
> I'm also seeing this having just attempted to update my repo from `nx@14.8.6` to `nx@15.8.7`. Everything works perfectly locally, but when CI (CircleCI) runs I'm getting no cache hits after the update, and the final unit test (using Jest) hangs indefinitely. I can't share my repo as it is proprietary, but if I can find more details or a resolution I'll share.

still having the same issue with the latest version of NX, and the only way to make it work on the CI is using --runInBand command
